---
title: 'Prostate Cancer RNASeq Analysis Between African Americans and Causasian Americans:'
author: "Flemming Wu"
---

```{r include=FALSE}
setwd("C:/Users/flemm/velvilab/P_Ca_Analysis_012723")
```

### Load libraries

```{r load libs, message = FALSE}
library(readxl)
library(openxlsx)
library(magrittr)
library(purrr)
library(tidyverse)
library(tibble)
library(stringr)
library(RUVSeq)
library(ggplot2)
library(RColorBrewer)
library(DESeq2)
library(EnsDb.Hsapiens.v86)
library(EnhancedVolcano)
library(clusterProfiler)
library(org.Hs.eg.db)
```

### Read in and process data

```{r message=FALSE}
metadata <- read_xlsx("./RNAseq_Samples_012723.xlsx")
data <- read_xlsx("./Data_012723.xlsx")
```

```{r}
dim(data) #58780 genes in 20 samples
```

```{r change rownames}
#head(data) ## gene id columns are redundant, make them the rownames and remove columns
data <- column_to_rownames(data, var = "Geneid...1")
data <- dplyr::select(data, -starts_with("Geneid"))
```

```{r change column names}
new.names <- metadata %>%
  mutate(KGP_Sample_ID = str_extract(KGP_Sample_ID, "00[:digit:]{2}") %>% str_replace("00[0]?", "RNA")) %>%
  mutate(`Tissue Type` = ifelse(`Tissue Type` == "Malignant", "M", "N"), Race = ifelse(Race == "Black", "B", "W")) %>%
  unite(id, c("Tissue Type", "Race", "MGC_Sample_ID")) %>%
  arrange(as.integer(str_extract(KGP_Sample_ID, "[:digit:]{1,2}"))) %>%
  dplyr::select(id) %>%
  as.list()

names(data) <- new.names$id
```

### Split data by comparisons

*reference level (control) in design matrix is the first level of the factor variable, which is the level with the lowest number or letter among the values provided*

```{r split by comparisons}
#head(table(rowMeans(data)), 20)
#head(table(rowSums(data)), 20) # 18,082 genes have 0 expression across all samples

## split data for all comparisons


### since the reference group in the design matrix is assigned by the level with the lowest number or letter among the values provided, I named the groups to match accordingly
## ex: mb_nb = Nonmalignant black (reference) vs nonmalignant black (comparison), since m comes before n


# Nonnmalignant black vs malignant black
mb_nb <- data %>% select_if(grepl("M_B", names(.)) | grepl("N_B", names(.)))
# Walignant white vs nonmalignant black
mw_nb <- data %>% select_if(grepl("M_W", names(.)) | grepl("N_B", names(.)))
# Nonmalignant black vs nonmalignant white
nb_nw <- data %>% select_if(grepl("N_B", names(.)) | grepl("N_W", names(.)))
# Malignant black vs malignant white
mb_mw <- data %>% select_if(grepl("M_B", names(.)) | grepl("M_W", names(.)))
# Malignant black vs nonmalignant white
mb_nw <- data %>% select_if(grepl("M_B", names(.)) | grepl("N_W", names(.)))
# Malignant white vs nonmalignant white
mw_nw <- data %>% select_if(grepl("M_W", names(.)) | grepl("N_W", names(.)))

## combine into list of datasets
all.dfs <- list("mb_nb" = mb_nb, "mw_nb" = mw_nb, "nb_nw" = nb_nw, "mb_mw" = mb_mw, "mb_nw" = mb_nw, "mw_nw" = mw_nw)
```

### Remove unwanted variation with `RUVg`

```{r}
# require more than 5 reads in at least 3 samples from each comparison
all.dfs.filt <- lapply(all.dfs, function(df) df[apply(df, 1, function(x) length(x[x>5])>=3), ])
```


```{r}
make_sets <- function(df) {
  x <- names(df) %>% str_extract("[MN]_[WB]") %>% as.factor()
  set <- newSeqExpressionSet(as.matrix(df), phenoData = data.frame(x, row.names = colnames(df)))
  return(set)
}

all.sets <- lapply(all.dfs.filt, function(x) make_sets(x))
```


```{r initial plots, fig.width=3, fig.height=2}
plot_sets <- function(set) {
  x <- set$x
  colors <- brewer.pal(3, "Set1")
  plotRLE(set, outline = FALSE, ylim = c(-4,4), col = colors[x])
  plotPCA(set, col = colors[x], cex = 0.7)
}

#lapply(all.sets, function(set) plot_sets(set))
```

#### Run between lane (upper quartile) normalization.

```{r between lane normalization, fig.width=3, fig.height=2}
all.sets.norm <- lapply(all.sets, function(set) betweenLaneNormalization(set, which = "upper"))

#lapply(all.sets.norm, function(set) plot_sets(set))
```


### Create empirical negative control genes and Run DE analysis with `edgeR`


```{r}
est_neg_ctrl <- function(set) {
  # Since no genes are known _a priori_ to be influenced by covariates of interest, use least significantly DE genes as "in-silico empirical" negative controls.
  x <- set$x
  design <- model.matrix(~x, data = pData(set))
  y <- DGEList(counts = counts(set), group = x)
  y <- calcNormFactors(y, method = "upperquartile")
  y <- estimateGLMCommonDisp(y, design)
  y <- estimateGLMTagwiseDisp(y, design)

  fit <- glmFit(y, design)
  lrt <- glmLRT(fit, coef = 2)

  top <- topTags(lrt, n = nrow(set))$table
  empirical <- rownames(set)[which(!(rownames(set) %in% rownames(top)[1:5000]))]
  
  # run RUVg
  newset <- RUVg(set, empirical, k = 1)
  return(newset)
}


run_glmLRT <- function(newset) {
  # Rerun edgeR with newly obtained negative controls
  x <- newset$x
  design <- model.matrix(~x + W_1, data = pData(newset))
  y <- DGEList(counts = counts(newset), group = x)
  y <- calcNormFactors(y, method = "upperquartile")
  y <- estimateGLMCommonDisp(y, design)
  y <- estimateGLMTagwiseDisp(y, design)

  fit <- glmFit(y, design)
  lrt <- glmLRT(fit, coef=2)
  return(lrt)
}


################



all.sets.norm.nc <- lapply(all.sets.norm, est_neg_ctrl)
all.res <- lapply(all.sets.norm.nc, run_glmLRT)

all.tables <- lapply(all.res, function(x) return(x$table))
```



```{r}
add_gene_sym <- function(df) {
  # remove "." from ensembl gene id
  rownames(df) <- sub("\\.[0-9]*$", "", rownames(df))
  # map ensembl geneid to gene symbol
  geneids <- ensembldb::select(x = EnsDb.Hsapiens.v86,  key=rownames(df), columns=c("SYMBOL"), keytype="GENEID")
  rownames(geneids) <- geneids$GENEID
  
  # merge mapping back to original dataframe
  df <- merge(x = df, y = geneids, by = "row.names", all = TRUE)
  column_to_rownames(df, var = "Row.names")
  
  return(df)
}

all.tables.sym <- lapply(all.tables, add_gene_sym)
```

```{r}
lapply(all.tables.sym, head)
```

```{r write results to excel}
## write results to excel workbook


## function to filter data based on pvalue cutoffs
get_sig_genes <- function(df, p_cutoff) {
  df <- df %>%
    column_to_rownames("Row.names") %>%
    dplyr::select(-c(GENEID)) %>%
    dplyr::filter(PValue < p_cutoff)
  return(df)
}

genes.05 <- lapply(all.tables.sym, get_sig_genes, 0.05)
genes.01 <- lapply(all.tables.sym, get_sig_genes, 0.01)
genes.001 <- lapply(all.tables.sym, get_sig_genes, 0.001)

if(!file.exists("gene_lists/genes_0.05.xlsx") & !file.exists("gene_lists/genes_0.01.xlsx") & !file.exists("gene_lists/genes_0.001.xlsx")) {
  write.xlsx(genes.05, file = "./gene_lists/genes_0.05.xlsx")
  write.xlsx(genes.01, file = "./gene_lists/genes_0.01.xlsx")
  write.xlsx(genes.001, file = "gene_lists/genes_0.001.xlsx")
}
```


## Pathway Analysis

```{r, eval=FALSE, include=FALSE}
#lapply(genes.05, nrow) ## too many genes (998 - 1927)
#lapply(genes.01, nrow) ## 0.01 p-value cutoff gives good amount of genes (301 - 856)
#lapply(genes.001, nrow) ## too few genes (75 - 336)


#############


## to create named gene list vector
#d <- genes.01$nb_mb %>%
#  arrange(desc(abs(logFC))) %>%
#  select(SYMBOL, logFC)

#geneList <- d[,2]
#names(geneList) <- d[,1]
#geneList

#enrichGO(gene = names(geneList), OrgDb = org.Hs.eg.db, keyType = "SYMBOL")
##

#res <- enrichGO(gene = rownames(genes.01$nb_mb), OrgDb = org.Hs.eg.db, keyType = "ENSEMBL")
res <- gseGO(gene = sort(geneList, decreasing = TRUE), OrgDb = org.Hs.eg.db, keyType = "SYMBOL")
res

res <- enrichKEGG(gene = genes.01$nb_mb$SYMBOL, organism = "hsa", keyType = "ncbi-geneid")
```


## Volcano Plots


```{r fig.height=6, fig.width=8}
lapply(names(all.tables.sym), function(x) EnhancedVolcano(all.tables.sym[[x]], x = "logFC", y = "PValue", lab = all.tables.sym[[x]]$SYMBOL,
                                                       title = x %>%
                                                         str_replace_all("n", "Non-Malignant ") %>%
                                                         str_replace_all("m", "Malignant ") %>%
                                                         str_replace_all("b", "African") %>%
                                                         str_replace_all("w", "Caucasian") %>%
                                                         str_replace_all("_", " -VS- "),
                                                       subtitle = "p-value cutoff: 0.05 | log2-fold change cutoff: 1", col = c("gray", "red2", "yellow1", "blue3")
                                                       ))
```



```{r fig.height=6}
#invisible(lapply(all.res, plotSmear, de.tags = "glmLRT", lowess = FALSE, main = "MA Plot"))

invisible(lapply(names(all.res), function(x) {plotSmear(all.res[[x]], main = x %>%
                                                         str_replace_all("n", "Non-Malignant ") %>%
                                                         str_replace_all("m", "Malignant ") %>%
                                                         str_replace_all("b", "African") %>%
                                                         str_replace_all("w", "Caucasian") %>%
                                                         str_replace_all("_", " -VS- "))}))
```





---





`DESeq2`

```{r Run DESeq2, message=FALSE}
run_DESeq <- function(set) {
  x <- set$x
  dds <- DESeqDataSetFromMatrix(countData = counts(set), colData = pData(set), design = ~W_1 + x)
  dds <- DESeq(dds)
  res <- results(dds)
  return(res)
}

all.deseq.res <- lapply(all.sets.norm.nc, run_DESeq)
```

```{r}
## convert to data frame
all.deseq.res <- lapply(all.deseq.res, as.data.frame)

## add gene symbols
all.deseq.res.sym <- lapply(all.deseq.res, add_gene_sym)
```


```{r, fig.height=5}
lapply(names(all.deseq.res.sym), function(x) EnhancedVolcano(all.deseq.res.sym[[x]], x = "log2FoldChange", y = "padj", lab = all.deseq.res.sym[[x]]$SYMBOL,
                                                       title = x %>%
                                                         str_replace_all("n", "Non-Malignant ") %>%
                                                         str_replace_all("m", "Malignant ") %>%
                                                         str_replace_all("b", "African") %>%
                                                         str_replace_all("w", "Caucasian") %>%
                                                         str_replace_all("_", " -VS- "),
                                                       subtitle = "p-value cutoff: 0.05 | log2-fold change cutoff: 1", col = c("gray", "red2", "yellow1", "blue3")
                                                       ))
```


