---
title: 'Prostate Cancer RNASeq Analysis Between African Americans and Causasian Americans:'
author: "Flemming Wu"
---

```{r include=FALSE}
if (Sys.info()["sysname"] == "Darwin") {setwd("~/Desktop/velvilab/P_Ca_Analysis")}
if (Sys.info()["sysname"] == "Windows") {setwd("C:/Users/flemm/velvilab/P_Ca_Analysis_012723")}
```

### Load libraries

```{r load libs, message = FALSE}
library(readxl)
library(openxlsx)
library(magrittr)
library(purrr)
library(tidyverse)
library(tibble)
library(stringr)
library(RUVSeq)
library(ggplot2)
library(RColorBrewer)
library(DESeq2)
library(EnsDb.Hsapiens.v86)
library(EnhancedVolcano)
library(clusterProfiler)
library(org.Hs.eg.db)
library(ggpubr)
library(msigdbr)
```

### Read in and process data

```{r message=FALSE}
metadata <- read_xlsx("./data/RNAseq_Samples_012723.xlsx")
data <- read_xlsx("./data/Data_012723.xlsx")
```

```{r}
dim(data) #58780 genes in 20 samples
```

```{r change rownames}
#head(data) ## gene id columns are redundant, make them the rownames and remove columns
data <- column_to_rownames(data, var = "Geneid...1")
data <- dplyr::select(data, -starts_with("Geneid"))
```

```{r change column names}
new.names <- metadata %>%
  mutate(KGP_Sample_ID = str_extract(KGP_Sample_ID, "00[:digit:]{2}") %>% str_replace("00[0]?", "RNA")) %>%
  mutate(`Tissue Type` = ifelse(`Tissue Type` == "Malignant", "M", "N"), Race = ifelse(Race == "Black", "B", "W")) %>%
  unite(id, c("Tissue Type", "Race", "MGC_Sample_ID")) %>%
  arrange(as.integer(str_extract(KGP_Sample_ID, "[:digit:]{1,2}"))) %>%
  dplyr::select(id) %>%
  as.list()

names(data) <- new.names$id
```

### Split data by comparisons

*reference level (control) in design matrix is the first level of the factor variable, which is the level with the lowest number or letter among the values provided*

```{r split by comparisons}
#head(table(rowMeans(data)), 20)
#head(table(rowSums(data)), 20) # 18,082 genes have 0 expression across all samples

## split data for all comparisons


### since the reference group in the design matrix is assigned by the level with the lowest number or letter among the values provided, I named the groups to match accordingly
## ex: mb_nb = Nonmalignant black (reference) vs nonmalignant black (comparison), since m comes before n


# Nonnmalignant black vs malignant black
mb_nb <- data %>% select_if(grepl("M_B", names(.)) | grepl("N_B", names(.)))
# Walignant white vs nonmalignant black
mw_nb <- data %>% select_if(grepl("M_W", names(.)) | grepl("N_B", names(.)))
# Nonmalignant black vs nonmalignant white
nb_nw <- data %>% select_if(grepl("N_B", names(.)) | grepl("N_W", names(.)))
# Malignant black vs malignant white
mb_mw <- data %>% select_if(grepl("M_B", names(.)) | grepl("M_W", names(.)))
# Malignant black vs nonmalignant white
mb_nw <- data %>% select_if(grepl("M_B", names(.)) | grepl("N_W", names(.)))
# Malignant white vs nonmalignant white
mw_nw <- data %>% select_if(grepl("M_W", names(.)) | grepl("N_W", names(.)))

## combine into list of datasets
all.dfs <- list("mb_nb" = mb_nb, "mw_nb" = mw_nb, "nb_nw" = nb_nw, "mb_mw" = mb_mw, "mb_nw" = mb_nw, "mw_nw" = mw_nw)
```

### Remove unwanted variation with `RUVg`

```{r}
# require more than 5 reads in at least 3 samples from each comparison
all.dfs.filt <- lapply(all.dfs, function(df) df[apply(df, 1, function(x) length(x[x>5])>=3), ])
```


```{r}
make_sets <- function(df) {
  x <- names(df) %>% str_extract("[MN]_[WB]") %>% as.factor()
  set <- newSeqExpressionSet(as.matrix(df), phenoData = data.frame(x, row.names = colnames(df)))
  return(set)
}

all.sets <- lapply(all.dfs.filt, function(x) make_sets(x))
```


```{r initial plots, fig.width=3, fig.height=2}
plot_sets <- function(set) {
  x <- set$x
  colors <- brewer.pal(3, "Set1")
  plotRLE(set, outline = FALSE, ylim = c(-4,4), col = colors[x])
  plotPCA(set, col = colors[x], cex = 0.7)
}

#lapply(all.sets, function(set) plot_sets(set))
```

#### Run between lane (upper quartile) normalization.

```{r between lane normalization, fig.width=3, fig.height=2}
all.sets.norm <- lapply(all.sets, function(set) betweenLaneNormalization(set, which = "upper"))

#lapply(all.sets.norm, function(set) plot_sets(set))
```


### Create empirical negative control genes and Run DE analysis with `edgeR`


```{r}
est_neg_ctrl <- function(set) {
  # Since no genes are known _a priori_ to be influenced by covariates of interest, use least significantly DE genes as "in-silico empirical" negative controls.
  x <- set$x
  design <- model.matrix(~x, data = pData(set))
  y <- DGEList(counts = counts(set), group = x)
  y <- calcNormFactors(y, method = "upperquartile")
  y <- estimateGLMCommonDisp(y, design)
  y <- estimateGLMTagwiseDisp(y, design)

  fit <- glmFit(y, design)
  lrt <- glmLRT(fit, coef = 2)

  top <- topTags(lrt, n = nrow(set))$table
  empirical <- rownames(set)[which(!(rownames(set) %in% rownames(top)[1:5000]))]
  
  # run RUVg
  newset <- RUVg(set, empirical, k = 1)
  return(newset)
}


run_glmLRT <- function(newset) {
  # Rerun edgeR with newly obtained negative controls
  x <- newset$x
  design <- model.matrix(~x + W_1, data = pData(newset))
  y <- DGEList(counts = counts(newset), group = x)
  y <- calcNormFactors(y, method = "upperquartile")
  y <- estimateGLMCommonDisp(y, design)
  y <- estimateGLMTagwiseDisp(y, design)

  fit <- glmFit(y, design)
  lrt <- glmLRT(fit, coef=2)
  return(lrt)
}


################



all.sets.norm.nc <- lapply(all.sets.norm, est_neg_ctrl)
all.res <- lapply(all.sets.norm.nc, run_glmLRT)
```


#### Filter gene lists based on: p-val= 0.05, p-val= 0.01, p-val= 0.001, FDR: 10%

```{r}
filterResults <- function(dgelrt, filt, thresh){
  topTags(dgelrt, sort.by = "p.value", n = nrow(dgelrt), adjust.method = "BH") %>%
  as.data.frame() %>%
  filter(if (filt == "pval") PValue < thresh else FDR < thresh)
}

genes.p05 <- lapply(all.res, filterResults, "pval", 0.05)
genes.p01 <- lapply(all.res, filterResults, "pval", 0.01)
genes.p001 <- lapply(all.res, filterResults, "pval", 0.001)
genes.fdr1 <- lapply(all.res, filterResults, "logfc", 0.1)
```

Add gene symbol to results

```{r}
addGeneSym <- function(df) {
  # remove "." from ensembl gene id
  rownames(df) <- sub("\\.[0-9]*$", "", rownames(df))
  # map ensembl geneid to gene symbol
  geneids <- ensembldb::select(x = EnsDb.Hsapiens.v86,  key=rownames(df), columns=c("SYMBOL"), keytype="GENEID")
  rownames(geneids) <- geneids$GENEID
  
  # merge mapping back to original dataframe
  df <- merge(x = df, y = geneids, by = "row.names", all = TRUE) %>%
    column_to_rownames(var = "Row.names") %>%
    relocate(SYMBOL, .before = logFC) %>%
    select(!starts_with("GENEID"))
  
  return(df)
}


genes.p05 <- lapply(genes.p05, addGeneSym)
genes.p01 <- lapply(genes.p01, addGeneSym)
genes.p001 <- lapply(genes.p001, addGeneSym)
genes.fdr1 <- lapply(genes.fdr1, addGeneSym)
```


Write results to excel workbook

```{r write results to excel}
if(!file.exists("gene_lists/genes_p05.xlsx") & !file.exists("gene_lists/genes_p01.xlsx") & !file.exists("gene_lists/genes_p001.xlsx") & !file.exists("gene_lists/genes_fdr1.xlsx")) {
  write.xlsx(genes.p05, file = "./gene_lists/genes_p05.xlsx", rowNames = TRUE)
  write.xlsx(genes.p01, file = "./gene_lists/genes_p01.xlsx", rowNames = TRUE)
  write.xlsx(genes.p001, file = "gene_lists/genes_p001.xlsx", rowNames = TRUE)
  write.xlsx(genes.fdr1, file = "gene_lists/genes_fdr1.xlsx", rowNames = TRUE)
}
```


## Gene Set Enrichment Analysis

GO and KEGG are the most frequently used for functional analysis. They are typically the first choice because of their long-standing curation and availability for a wide range of species.

In conclusion, analysing up- and downregulated genes separately is more powerful than analysing all of the DE genes together.

**IT IS BETTER TO RUN GSEA ON ALL GENES, NOT ONLY OVER/UNDER EXPRESSED**
**GSEA RANKS GENES BASED ON CERTAIN VALUE YOU PROVIDE**
Then, this is VERY important, the result it is not that a specific pathway is up- or down-regulated, but the fact that the pathway is affected in some way by the condition that you re studying.
https://www.biostars.org/p/449956/

The GSEA algorithm does not filter the expression dataset and generally does not benefit from your filtering of the expression dataset. During the analysis, genes that are poorly expressed or that have low variance across the dataset populate the middle of the ranked gene list and the use of a weighted statistic ensures that they do not contribute to a positive enrichment score. By removing such genes from your dataset, you may actually reduce the power of the statistic.

First I will perform gene set enrichment analysis on GO terms.

```{r}
### DELETE, DO NOT FILTER GENE LISTS

## prepare gene lists to input to gsea
## clusterProfiler requires a named list of logFC for input

## p-value 0.05
geneList.p05 <- lapply(names(genes.p05), function(x){
  d <- genes.p05[[x]] %>%
    arrange(desc(logFC)) %>%
    select(logFC)
  geneList <- d[,1]
  names(geneList) <- rownames(d)
  return(geneList)
})
names(geneList.p05) <- names(genes.p05)


## p-value 0.01
geneList.p01 <- lapply(names(genes.p01), function(x){
  d <- genes.p01[[x]] %>%
    arrange(desc(logFC)) %>%
    select(logFC)
  geneList <- d[,1]
  names(geneList) <- rownames(d)
  return(geneList)
})
names(geneList.p01) <- names(genes.p01)


## p-value 0.001
geneList.p001 <- lapply(names(genes.p001), function(x){
  d <- genes.p001[[x]] %>%
    arrange(desc(logFC)) %>%
    select(logFC)
  geneList <- d[,1]
  names(geneList) <- rownames(d)
  return(geneList)
})
names(geneList.p001) <- names(genes.p001)



## false discovery rate 10%
geneList.fdr1 <- lapply(names(genes.fdr1), function(x){
  d <- genes.fdr1[[x]] %>%
    arrange(desc(logFC)) %>%
    select(logFC)
  geneList <- d[,1]
  names(geneList) <- rownames(d)
  return(geneList)
})
names(geneList.fdr1) <- names(genes.fdr1)
```

### Perform GO enrichmennt analysis


```{r}
## functions to run gseGO with different params
rungseGOstd <- function(geneList){
  return(gseGO(geneList, ont = "ALL", keyType = "ENSEMBL", OrgDb = org.Hs.eg.db, nPermSimple = 10000, eps = 1e-300, scoreType = "std"))
}

rungseGOup <- function(geneList){
  return(gseGO(geneList, ont = "ALL", keyType = "ENSEMBL", OrgDb = org.Hs.eg.db, nPermSimple = 10000, eps = 0, scoreType = "pos"))
}

rungseGOdn <- function(geneList){
  return(gseGO(geneList, ont = "ALL", keyType = "ENSEMBL", OrgDb = org.Hs.eg.db, nPermSimple = 10000, eps = 0, scoreType = "neg"))
}
```

```{r}
## calculate logFC * pvalue for ranking statistic
gsea.input <- lapply(all.res, function(x) topTags(x, n = nrow(x)) %>% as.data.frame %>% mutate(stat = logFC*PValue) %>% arrange(-stat))

## created named list
gsea.input$mb_nb %>% rownames_to_column()
```


## Perform gene set enrichment analyses on only up or only downregulated genes

```{r, eval=FALSE, include = FALSE}

## separate list of up and down regulated genes
geneList.p05up <- lapply(geneList.p05, function(x) x[x > 0])
geneList.p05dn <- lapply(geneList.p05, function(x) x[x < 0])
geneList.p01up <- lapply(geneList.p01, function(x) x[x > 0])
geneList.p01dn <- lapply(geneList.p01, function(x) x[x > 0])
geneList.p001up <- lapply(geneList.p001, function(x) x[x > 0])
geneList.p001dn <- lapply(geneList.p001, function(x) x[x < 0])
geneList.fdr1up <- lapply(geneList.fdr1, function(x) x[x > 0])
geneList.fdr1dn <- lapply(geneList.fdr1, function(x) x[x < 0])


ego.p05up <- lapply(geneList.p05up, rungseGOup)
ego.p05dn <- lapply(geneList.p05dn, rungseGOdn)

ego.p05up.std <- lapply(geneList.p01up, rungseGOstd)

lapply(ego.p05up, function(x) x@result) # returns 0 significantly enriched pathways
lapply(ego.p05dn, function(x) x@result) # returns only 1 significantly enriched pathways
```


## trying `fgsea`

```{r}
#msigdbr_collections()
```


```{r}
library(fgsea)
genes.fdr1$nb_nw %>%
  select(SYMBOL, LR) %>%
  deframe()

msigdbr_df <- msigdbr(species = "Homo sapiens", category = "C5")
msigdbr_list <- split(x = msigdbr_df$ensembl_gene, f = msigdbr_df$gs_name)

fgsea(pathways = msigdbr_list, 
      stats = genes.fdr1$nb_nw %>%
        rownames_to_column("ENSEMBL") %>%
        select(ENSEMBL, PValue) %>%
        arrange(PValue) %>%
        deframe(), minSize = 10, maxSize = 500) %>%
  arrange(-padj)
```





## Volcano Plots


```{r fig.height=6, fig.width=8}
DoVolcanoPlot <- function(x){
  EnhancedVolcano(all.tables.sym[[x]], 
                  x = "logFC", 
                  y = "PValue", 
                  lab = all.tables.sym[[x]]$SYMBOL,
                  title = x %>%
                    str_replace_all("n", "Non-Malignant ") %>%
                    str_replace_all("m", "Malignant ") %>%
                    str_replace_all("b", "African") %>%
                    str_replace_all("w", "Caucasian") %>%
                    str_replace_all("_", " -VS- "),
                  subtitle = "p-value cutoff: 0.05 | log2-fold change cutoff: 1", 
                  col = c("gray", "#4bd0fe", "red1", "#fef84b")
  )
}

volplots <- lapply(names(all.tables.sym), DoVolcanoPlot)
names(volplots) <- names(all.tables.sym)
```

Save volcano plots
```{r}
#lapply(names(volplots), function(x) ggsave(filename=paste0("./plots/vol_", x,".jpeg"),
#                                          plot=volplots[[x]], dpi = 800, width = 9, 
#                                          height = 11, units = "in"))
```


## MA Plots

Typically, lower mean expression values will have more variability in log fold-change than the higher expression value. This results in a fanning effect of the data points as the graph moves from right to left.


```{r MA Plots, fig.height=6, fig.width=8}
DoMAPlot <- function(x){
  data <- all.tables.sym[[x]] %>%
    rename(log2FoldChange = logFC, baseMeanLog2 = logCPM, padj = PValue)
  genenames <- all.tables.sym[[x]]$SYMBOL %>% as.vector()
  ggmaplot(data = data, 
           main = x %>%
             str_replace_all("n", "Non-Malignant ") %>%
             str_replace_all("m", "Malignant ") %>%
             str_replace_all("b", "African") %>%
             str_replace_all("w", "Caucasian") %>%
             str_replace_all("_", " -VS- "),
           fdr = 0.05, fc = 2, size = 0.4,
           palette = c("#7AFF96", "#FF7A7A", "darkgray"),
           genenames = genenames,
           legend = "top", top = 20,
           select.top.method = "padj",
           font.label = c("bold", 9), label.rectangle = FALSE,
           font.legend = "bold",
           font.main = "bold",
           xlab = "Average logCPM",
           ggtheme = ggplot2::theme_minimal()
           )
}

maplots <- lapply(names(all.tables.sym), DoMAPlot)
names(maplots) <- names(all.tables.sym)
```

Save MA plots
```{r}
#lapply(names(maplots), function(x) ggsave(filename=paste0("./plots/ma_", x,".jpeg"),
#                                          plot=maplots[[x]], dpi = 800, width = 10, 
#                                         height = 7, units = "in"))
```


---




```{r Run DESeq2, message=FALSE, eval=FALSE, include=FALSE}
run_DESeq <- function(set) {
  x <- set$x
  dds <- DESeqDataSetFromMatrix(countData = counts(set), colData = pData(set), design = ~W_1 + x)
  dds <- DESeq(dds)
  res <- results(dds)
  return(res)
}

all.deseq.res <- lapply(all.sets.norm.nc, run_DESeq)

## convert to data frame
all.deseq.res <- lapply(all.deseq.res, as.data.frame)

## add gene symbols
all.deseq.res.sym <- lapply(all.deseq.res, add_gene_sym)

lapply(names(all.deseq.res.sym), function(x) EnhancedVolcano(all.deseq.res.sym[[x]], x = "log2FoldChange", y = "padj", lab = all.deseq.res.sym[[x]]$SYMBOL,
                                                       title = x %>%
                                                         str_replace_all("n", "Non-Malignant ") %>%
                                                         str_replace_all("m", "Malignant ") %>%
                                                         str_replace_all("b", "African") %>%
                                                         str_replace_all("w", "Caucasian") %>%
                                                         str_replace_all("_", " -VS- "),
                                                       subtitle = "p-value cutoff: 0.05 | log2-fold change cutoff: 1", col = c("gray", "red2", "yellow1", "blue3")
                                                       ))
```


