---
title: "Prostate Cancer Secondary Analysis With Validation Cohort"
author: "Flemming Wu"
subtitle: "Using `DESeq2` For DE Analysis"
---

```{r setup, include=F}
if (Sys.info()["sysname"] == "Windows") {
  knitr::opts_chunk$set(root.dir="~/velvilab/P_Ca_Secondary")
} else if (Sys.info()["sysname"] == "Darwin") {
  knitr::opts_chunk$set(rroot.dir="~/Desktop/velvilab/P_Ca_Analysis")
}
knitr::opts_chunk$set(fig.width=6, fig.height=4)
```

Load in required libraries

```{r load libs, message=F}
library(tidyverse)
library(RUVSeq)
library(readxl)
library(RColorBrewer)
library(EnsDb.Hsapiens.v86)
library(org.Hs.eg.db)
library(EnhancedVolcano)
library(ggpubr)
library(clusterProfiler)
library(enrichplot)
library(DESeq2)
library(patchwork)
```

## Load in and prepare data

Load in excel file to map samples to type

```{r}
key <- read_xlsx("./Manifest/P_Ca_sec_manifest_030923.xlsx")
head(key)
```

Read in one count table

```{r read in one file}
test.file <- read.table(
  file="./Count_Table/ENRIQUERNA_0001_01_PROSTATE_SRR10238298_C1_NEBKAP_A00001.proj.Aligned.out.sorted.md.bam.htSeqCounts"
)
head(test.file)
nrow(test.file)
```

Read all count tables into R and prepare genes x samples count matrix

```{r read in counts from Count Table directory}
## I have checked, and all files should have same the same genes listed in the same order

# create new df to hold count matrix
df <- as.data.frame(matrix(0, ncol = 0, nrow = nrow(test.file)))
for (i in list.files(path="./Count_Table")) {
  name <- str_extract(i, "SRR[0-9]{8}")
  df[[name]] <- read.table(file=paste0("./Count_Table/",i)) %>% pull(V2) # read in vector of counts to new column
}

# assign gene names to the row names
row.names(df) <- test.file$V1 

# assign African American or Caucasian American to sample names
names(df) <- ifelse(names(df) %in% key$`African American`, paste(names(df), "Afr", sep="_"), paste(names(df), "Cau", sep="_"))
```

```{r view prepared dataset}
#dim(df) # 58,785 genes and 59 samples
head(df)[,1:5]
```

Filter out lowly expressed genes

```{r filtering}
# filter criteria:
## need a read count of more than 5 in at least 15 samples (about 25% of the samples)
df_unfiltered <- df # save unfiltered dataframe as df_unfiltered
df <- df[apply(df, 1, function(x) length(x[x>5])>=15),]
```

Left with 27,786 genes after the filtering


### Exploratory data analysis

Create expression set object from count matrix

```{r create expression set object}
x <- names(df) %>% str_extract("Afr$|Cau$") %>% as.factor()
set <- newSeqExpressionSet(as.matrix(df), phenoData = data.frame(x, row.names = colnames(df)))
set
```

```{r plot RLE and PCA prior to normalization, fig.width=12, fig.height=6}
colors <- brewer.pal(3, "Set1")
par(mfrow=c(1,2))
plotRLE(set, outline = FALSE, ylim = c(-4,4), col = colors[x])
plotPCA(set, col = colors[x], cex = 0.7)
mtext("Before Normalization", side = 3, line = -1, outer = TRUE)
```

### Initial data normalization

```{r upper quartile normalization, fig.width=12, fig.height=6}
set <- betweenLaneNormalization(set, which = "upper")
par(mfrow=c(1,2))
plotRLE(set, outline = FALSE, ylim = c(-4,4), col = colors[x])
plotPCA(set, col = colors[x], cex = 0.7)
mtext("After Normalization", side = 3, line = -1, outer = TRUE)
```

### Remove covariates that are not of interest in comparison (AA vs CA)

#### `RUVSeq` offers 3 methods for removing unwanted variation in RNASeq analysis

1. `RUVg`: Estimating the factors of unwanted variation using control genes

2. `RUVs`: Estimating the factors of unwanted variation using replicate samples

3. `RUVr`: Estimating the factors of unwanted variation using residuals

I will try all three methods and evaluate which normalization method separates African American samples and Caucasian samples *best* by visualizing the separation on the PCA space.


#### `RUVg`

Obtain least significantly DE genes (based on first-pass DE analysis before) as negative control for removing unwanted variation

```{r run first pass DEG analysis}
design.matrix <- model.matrix(~x, data=pData(set)) # create design matrix

y <- DGEList(counts=counts(set), group=x) # create DGEList object from table of counts and group indicator

y <- calcNormFactors(y, method="upperquartile") # calculate scaling factors to convert raw library sizes into effective library sizes

y <- estimateGLMCommonDisp(y, design.matrix) # Estimates a common negative binomial dispersion parameter for a DGE dataset

y <- estimateGLMTagwiseDisp(y, design.matrix) # compute empirical Bayes estimate of negative binomial dispersion parameter for each tag

fit <- glmFit(y, design.matrix) # Fit a negative binomial generalized log-linear model to the read counts for each gene and conduct genewise statistical tests
lrt <- glmLRT(fit, coef=2)

top <- topTags(lrt, n=nrow(set))$table
neg.ctrl <- rownames(set)[which(!(rownames(set) %in% rownames(top)[1:5000]))] ## all but the top 5000 genes are negative controls
```

Now, `RUVSeq` asks to specify parameter `k`, which equals the number of factors of unwanted variation to be estimated from the data. I will vary the values of k from 1-9 and select which one results in best separation between the two groups

```{r vary k for RUVg, fig.width=15, fig.height=15}
par(mfrow=c(3,3))
for (i in 1:9) {
  set2 <- RUVg(set, neg.ctrl, k=i)
  plotPCA(set2, col=colors[x], cex=0.7)
}
```

k = 6 can be argued to give the best separation between groups, but it is possible the other two methods may give better results


#### `RUVs`

One can use the `RUVs` method to estimate the factors of unwanted variation using replicate/negative control samples for which the covariates of interest are constant

First, we need to construct a matrix specifying the replicates

```{r construct replicates matrix}
differences <- makeGroups(x)
differences
```

Although in principle one still needs control genes for the estimation of the factors of unwanted variation, we found that `RUVs` is robust to that choice and that using all the genes works well in practice

```{r vary k for RUVs, fig.width=15, fig.height=15}
par(mfrow=c(3,3))
for (i in 1:9) {
  set3 <- RUVs(set, rownames(set), k=i, differences)
  plotPCA(set3, col=colors[x], cex=0.7)
}
```

For the `RUVs` normalization, it seems that k = 9 gives the best separation between groups


### `RUVr`

The `RUVr` method considers the residuals (e.g., deviance residuals) from a first-pass GLM regression of the counts on the covariates of interest

First, we need to compute the residuals from the GLM fit, without `RUVg` normalization, but possibly after normalization using a method such as upper-quartile normalization

```{r get deviance residuals}
design <- model.matrix(~x, data=pData(set))
y <- DGEList(counts=counts(set), group=x)
y <- calcNormFactors(y, method="upperquartile")
y <- estimateGLMCommonDisp(y, design)
y <- estimateGLMTagwiseDisp(y, design)
fit <- glmFit(y, design)
res <- residuals(fit, type="deviance")
```

```{r vary k for RUVr, fig.width=15, fig.height=15}
par(mfrow=c(3,3))
for (i in 1:9) {
  set4 <- RUVr(set, rownames(set), k=i, res) # Again, we can use all the genes to estimate the factors of unwanted variation
  plotPCA(set4, col=colors[x], cex=0.7)
}
```

Using k = 9 from `RUVr` gives the clearest separation between groups on the first principal component from all 3 `RUVSeq` methods. I will use this normalization method for downstream analysis.

Remove unwanted variation with pearson residuals.

```{r RUVg, fig.width=12, fig.height=6}
set2 <- RUVr(set, rownames(set), k=9, res)
par(mfrow=c(1,2))
plotRLE(set2, outline = FALSE, ylim = c(-4,4), col = colors[x])
plotPCA(set2, col = colors[x], cex = 0.7)
mtext("After normalizing with deviance residuals", side = 3, line = -1, outer = TRUE)
```

### Run final DEG analysis with `DESeq2`

```{r run DESeq2, message=FALSE}
dds <- DESeqDataSetFromMatrix(countData = counts(set2), colData = pData(set2), design = ~W_1 + x)
dds <- DESeq(dds)
res <- results(dds)
res <- na.omit(res)
# baseMean is the average of the normalized count values, dividing by size factors, taken over all samples
res <- res[res$baseMean > 50,] # filter out lowly expressed genes to reduce noise
```

```{r view DESeq2 results}
#res
df.res <- as.data.frame(res) %>% arrange(pvalue)
head(df.res, n = 10)
```


```{r add gene symbol to results}
addGeneSym <- function(df) {
  # remove "." from ensembl gene id
  rownames(df) <- sub("\\.[0-9]*$", "", rownames(df))
  # map ensembl geneid to gene symbol
  geneids <- ensembldb::select(x = EnsDb.Hsapiens.v86,  key=rownames(df), columns=c("SYMBOL"), keytype="GENEID")
  rownames(geneids) <- geneids$GENEID
  
  # merge mapping back to original dataframe
  df <- merge(x = df, y = geneids, by = "row.names", all = TRUE) %>%
    column_to_rownames(var = "Row.names") %>%
    relocate(SYMBOL, .before = baseMean) %>%
    dplyr::select(!starts_with("GENEID"))
  
  return(df)
}

df.res.sym <- addGeneSym(df.res)
df.res.sym <- arrange(df.res.sym, pvalue)
```

##### Positive logFC means upregulated in Caucasian Americans compared to African Americans and vice versa for negative logFC

```{r view DESeq2 results with gene symbol}

head(df.res.sym, n = 20) 
```


### Visualization of significant genes

Volcano Plot

```{r volcano plot, fig.width=10, fig.height=9}
EnhancedVolcano(df.res.sym, x = "log2FoldChange", y = "pvalue", lab = df.res.sym$SYMBOL,
                title = "Comparison: African American x Caucasian American - DESeq2",
                subtitle = "p-value cutoff: 0.05 | log2-fold change cutoff: 1",
                col = c("gray", "#4bd0fe", "red1", "#fef84b")
)
```

```{r MA plot,fig.width=12, fig.height=7}
to.plot <- df.res.sym %>%
  mutate(baseMean = log2(baseMean)) %>%
  dplyr::rename(baseMeanLog2 = baseMean)
  
ggmaplot(data = to.plot, 
           main = "Comparison: African American x Caucasian American - DESeq2",
           fdr = 0.1, fc = 2, size = 0.4,
           palette = c("#7AFF96", "#FF7A7A", "darkgray"),
           genenames = to.plot$SYMBOL,
           legend = "top", top = 30,
           select.top.method = "padj",
           font.label = c("bold", 9), label.rectangle = FALSE,
           font.legend = "bold",
           font.main = "bold",
           xlab = "log2 Base Mean",
           ggtheme = ggplot2::theme_minimal()
           )
```

### Gene set enrichment analysis

GSEA requires a *ranked* list of all genes. There is no need to separate up and downregulated genes, or filter the gene list, as the ranking should place unsignificant genes in the middle of the ranked list.

However, there is no set rule for ranking the list of genes. The method that I will use to rank genes is to multiply the log fold-change by the -log10 p-value. This ranking allows one to take into account the *change* in gene expression between the two groups, as well as the significance of the gene.

```{r prepare ranked gene list}
ranked.gene.list <- df.res %>%
  mutate(rank = log2FoldChange * -log10(pvalue)) %>% ## here I rank by multiplying logFC by -log10(pvalue) to give more weight to more significant genes
  rownames_to_column(var = "ENSEMBL") %>% 
  mutate(ENSEMBL = str_replace(ENSEMBL, ".[0-9]{1,2}$", "")) %>% 
  dplyr::select(ENSEMBL, stat) %>%
  arrange(-stat) %>%
  tibble::deframe()
```

```{r go gsea}
gsea.go <- gseGO(geneList = ranked.gene.list, OrgDb = org.Hs.eg.db, keyType = "ENSEMBL", minGSSize = 10,
      maxGSSize = 500, nPermSimple = 10000, pvalueCutoff = 0.05, eps = 1e-300, verbose = F)

#head(gsea.go)
```

GSEA plot for top enriched pathway

```{r fig.width=10, fig.height=10}
gseaplot2(gsea.go, geneSetID = 1, title = gsea.go$Description[1], pvalue_table = TRUE)
```

```{r include=FALSE, eval=FALSE}
gseaplot(gsea.go, geneSetID = 1, title = gsea.go$Description[1])
```


GSEA plot for top 5 enriched pathways

```{r fig.width=10, fig.height=10}
gseaplot2(gsea.go, geneSetID = 1:5, pvalue_table = TRUE, title = "GSEA plot - Top 5 Pathways in Caucasian Americans")
```

```{r fig.height=12, fig.width=10}
dotplot(gsea.go, showCategory = 30) + ggtitle("GSEA Identified Top 30 Upregulated GO Pathways in Caucasian Americans")
```

### Overrepresentation analysis

GO Enrichment analysis with filtered gene lists

split gene lists into up and downregulated filtered by pvalue cutoff

```{r split gene lists into up and downregulated filtered by pvalue cutoffs}
p.05.up <- filter(df.res.sym, (pvalue < 0.05) & (log2FoldChange > 0))
p.05.dn <- filter(df.res.sym, (pvalue < 0.05) & (log2FoldChange < 0))

p.01.up <- filter(df.res.sym, (pvalue < 0.01) & (log2FoldChange > 0))
p.01.dn <- filter(df.res.sym, (pvalue < 0.01) & (log2FoldChange < 0))

p.001.up <- filter(df.res.sym, (pvalue < 0.001) & (log2FoldChange > 0))
p.001.dn <- filter(df.res.sym, (pvalue < 0.001) & (log2FoldChange < 0))

fdr.1.up <- filter(df.res.sym, (padj < 0.1) & (log2FoldChange > 0))
fdr.1.dn <- filter(df.res.sym, (padj < 0.1) & (log2FoldChange < 0))
```

```{r create list of filtered results}
filtered.res <- list("p.05.up" = p.05.up, "p.05.dn" = p.05.dn,
                     "p.01.up" = p.01.up, "p.01.dn" = p.01.dn,
                     "p.001.up" = p.001.up, "p.001.dn" = p.001.dn,
                     "fdr.1.up" = fdr.1.up, "fdr.1.dn" = fdr.1.dn)
```

```{r GO enrichment, message=F}
run_enrichGO <- function(x) {
  return(enrichGO(
    gene = AnnotationDbi::select(org.Hs.eg.db, 
                                      keys = rownames(x), 
                                      columns='ENTREZID', 
                                      keytype='ENSEMBL') %>% pull(ENTREZID), 
    ont = "BP", 
    OrgDb = org.Hs.eg.db, 
    readable = TRUE
  ))
}
ego.res <- lapply(filtered.res, run_enrichGO)
names(ego.res) <- names(filtered.res)

## add pairwise termism (clustering of enriched terms) to enrichment results
ego.res <- lapply(ego.res, pairwise_termsim)
```

```{r fig.width=15, fig.height=12}
treeplot(ego.res$fdr.1.up) + labs(title = "Upregulated Caucasian American GO: BP Pathways Clustered By Term Similarity", subtitle = "Sig. Genes Cutoff Used: FDR 10%")
```

```{r fig.width=15, fig.height=12}
treeplot(ego.res$fdr.1.dn) + labs(title = "Upregulated African American GO: BP Pathways Clustered By Term Similarity", subtitle = "Sig. Genes Cutoff Used: FDR 10%")
```

```{r barplots of top 20 pathways, fig.width=16, fig.height=14}
wrap_plots(
  barplot(ego.res$fdr.1.up, showCategory=20) + ggtitle("Upreg in Caucasian American"),
  barplot(ego.res$fdr.1.dn, showCategory=20) + ggtitle("Upreg in African American")
)
```

```{r fig.height=2, fig.width=10, eval=F, include=F}
heatplot(ego.res$fdr.1.up, showCategory = 5)
```



#### KEGG Enrichment analysis with filtered gene lists

```{r, eval=F, include=F}
mkk <- enrichMKEGG(gene = bitr(rownames(filtered.res$p.05.up), fromType = "GENEID",
                               toType = "ENTREZID", OrgDb = EnsDb.Hsapiens.v86) %>% pull(ENTREZID),
                   keyType = 'ncbi-geneid',
                   organism = 'hsa',
                   pvalueCutoff = 1,
                   qvalueCutoff = 1)

head(mkk)
mkk
```




